### PR TITLE
Parse dependency if only name is present

### LINKED ISSUE #2229: Go parser cannot parse go.mod and go.sum files without a namespace
### Description

> Please leave a brief description of the bug or feature request:

Normally an entry to e.g. `go.sum` looks something like this:
```
github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
```
The Go parser parses `github.com/BurntSushi` into `namespace` and `toml` into `name`.

I have this entry:
```
go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=
```
Basically the `namespace` portion is missing.

For my concrete issue this is the corresponding source code: https://github.com/nexB/scancode-toolkit/blob/v3.2.1rc2/src/packagedcode/go_mod.py#L253-L254, but `namespace` is used through out the Go parser.

My proposal is to remove the `namespace` portion and parse everything into `name`, since I do not see in the source that `namespace` is used standalone somewhere.

### How To Reproduce

> Tell us how to reproduce the issue. 

### System configuration

> For bug reports, it really helps us to know:

* What OS are you running on? (Windows/MacOS/Linux) Ubuntu 20.04
* What version of scancode-toolkit was used to generate the scan file? v3.2.1rc2
* What installation method was used to install/run scancode? (pip/source download/other) `./configure`

### PR DESCRIPTION
**Pull request by rpotter12**
Fixes #2229

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

File: src/packagedcode/go_mod.py [hunk 1]
```diff
@@ -250,15 +250,19 @@ def parse_gomod(location):
...
     r'\s*'
     r'h1:(?P<checksum>[^\s]*)\s*'
 ).match
 
+name_parser = re.compile(
+    r'(?P<namespace>[^\s]*)'
+    r'\/'
+    r'(?P<name>[^\s]*)'
+).match
+
```
Comment (pombredanne) [inline src/packagedcode/go_mod.py:265]: I think I get your approach: get a `namespace/name` first and if it contains a `/` further split it afterwards below, right?
If so, let me suggest a slightly simpler approach for the second part below which would mean you do not need that second regex above

File: src/packagedcode/go_mod.py [hunk 2]
```diff
@@ -273,9 +277,16 @@ def parse_gosum(location):
         line = line.replace('/go.mod', '')
         parsed_dep = get_dependency(line)
 
+        name = parsed_dep.group('name')
+        namespace = ''
+        if '/' in name:
+            dep_name = name_parser(name)
+            namespace = dep_name.group('namespace')
+            name = dep_name.group('name')
+
```
Comment (pombredanne) [inline src/packagedcode/go_mod.py:286]: here is my suggestion:
```
ns_name = parsed_dep.group('name')
namespace, _, name = ns_name.rpartition('/')
```
the partition and rpartition functions are your friends most often over regex when splitting strings:
```
>>> 'foo'.rpartition('/')
('', '', 'foo')
>>> 'foo/bar'.rpartition('/')
('foo', '/', 'bar')
>>> 'foo/bar/baz'.rpartition('/')
('foo/bar', '/', 'baz')
```

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you! See my comments inline. Also do you tests the same thing for a go.mod such as there https://github.com/nexB/scancode-toolkit/issues/2229#issuecomment-692835210 ?

**Comment by pombredanne** [comment]
@jmastr ping

File: src/packagedcode/go_mod.py (hunk 2 above)
Comment (rpotter12) [reply src/packagedcode/go_mod.py:286]: This will be more better than mine. Thanks :)

**Comment by rpotter12** [comment]
@pombredanne I have updated the scripts. Please review it again.:)

File: src/packagedcode/go_mod.py [hunk 3]
```diff
@@ -250,15 +250,19 @@ def parse_gomod(location):
...
-    r'(?P<namespace>[^\s]*)'
-    r'\/'
     r'(?P<name>[^\s]*)'
     r'\s*'
     r'(?P<version>[^\s]*)'
     r'\s*'
     r'h1:(?P<checksum>[^\s]*)\s*'
 ).match
 
+name_parser = re.compile(
```
Comment (pombredanne) [inline src/packagedcode/go_mod.py:260]: I think this whole block is no longer needed, is it?

File: src/packagedcode/go_mod.py [hunk 4]
```diff
@@ -250,15 +250,19 @@ def parse_gomod(location):
 # Regex expressions to parse go.sum file dependency
 # dep example: github.com/BurntSushi/toml v0.3.1 h1:WXkYY....
 get_dependency = re.compile(
-    r'(?P<namespace>[^\s]*)'
-    r'\/'
     r'(?P<name>[^\s]*)'
```
Comment (pombredanne) [inline src/packagedcode/go_mod.py:253]: what if you named this `ns_name` or `namespace_name` rather than `name` since this is what it is now?

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! I added some more comments. Could you also add tests with a go.mod?

**Comment by rpotter12** [comment]
@pombredanne Please review it again :)

**Comment by jmastr** [comment]
@rpotter12 Currently on my phone, but is your patch „only“ for go.sum? Can’t this happen in go.mod too? Sry, if I am missing something here.

**Comment by rpotter12** [comment]
@pombredanne @jmastr 
1. https://github.com/census-instrumentation/opencensus-service/blob/57e037c599b75fe2dddf89cb9ada67bf4c8de9fc/go.mod
2. https://github.com/census-instrumentation/opencensus-service/blob/57e037c599b75fe2dddf89cb9ada67bf4c8de9fc/go.sum

I don't think we have to solve this because `go.opencensus.io` is not used like `go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=`

What do you say this??

**Comment by jmastr** [comment]
> @pombredanne @jmastr
> 
>     1. https://github.com/census-instrumentation/opencensus-service/blob/57e037c599b75fe2dddf89cb9ada67bf4c8de9fc/go.mod
> 
>     2. https://github.com/census-instrumentation/opencensus-service/blob/57e037c599b75fe2dddf89cb9ada67bf4c8de9fc/go.sum
> 
> 
> I don't think we have to solve this because `go.opencensus.io` is not used like `go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=` as mentioned in above examples.
> 
> What do you say this??

@rpotter12 I have a mod file from google like this:
```
go.opencensus.io v0.22.4
```
https://github.com/googleapis/google-cloud-go/blob/v0.66.0/go.mod#L14

**Comment by rpotter12** [comment]
@jmastr Okay then. I have solved this problem. You can review it :)

**Comment by jmastr** [comment]
LGTM

**Comment by rpotter12** [comment]
@pombredanne Please review this too :)

**Review by pombredanne — CHANGES_REQUESTED** [review]
@rpotter12 Thank you! I have a few comments on the regex that I had not noticed before.

File: src/packagedcode/go_mod.py [hunk 5]
```diff
@@ -123,17 +123,15 @@ def purl(self, include_version=True):
...
 # Regex expressions to parse different types of go.mod file dependency
 parse_module_name = re.compile(
-    r'^module\s*'
-    r'(?P<namespace>[^\s]*)'
-    r'\/'
-    r'(?P<name>[^\s]*)\s*'
+    r'(?P<type>[^\s]*)'
+    r'(\s)*'
+    r'(?P<ns_name>[^\s]*)'
+    r'\s*(?P<version>[^\s]*)\s*'
```
[the remark this answers is not in the document]
Comment (rpotter12) [reply src/packagedcode/go_mod.py:129]: This suggestion doesn't work. I don't know why.

[the remark this answers is not in the document]
Comment (rpotter12) [reply src/packagedcode/go_mod.py:129]: Ah....we need to add `r'\s?' in line 129`

File: src/packagedcode/go_mod.py [hunk 6]
```diff
@@ -165,19 +163,25 @@ def parse_gomod(location):
 
         parsed_module_name = parse_module_name(line)
         if parsed_module_name:
-            gomods.name = parsed_module_name.group('name')
-            gomods.namespace = parsed_module_name.group('namespace')
+            ns_name = parsed_module_name.group('ns_name')
+            namespace, _, name = ns_name.rpartition('/')
+
+        if 'module' in line:
```
[the remark this answers is not in the document]
Comment (rpotter12) [reply src/packagedcode/go_mod.py:169]: @pombredanne There are three types of lines:
1. module github.com/alecthomas/participle
2. require github.com/davecgh/go-spew v1.1.1
3. exclude github.com/davecgh/go-spew v1.1.1

so `parsed_module_name` parse all the three types of lines.

File: src/packagedcode/go_mod.py [hunk 7]
```diff
@@ -123,17 +123,15 @@ def purl(self, include_version=True):
 
 # Regex expressions to parse different types of go.mod file dependency
 parse_module_name = re.compile(
-    r'^module\s*'
-    r'(?P<namespace>[^\s]*)'
-    r'\/'
-    r'(?P<name>[^\s]*)\s*'
+    r'(?P<type>[^\s]*)'
+    r'(\s)*'
```
[the remark this answers is not in the document]
Comment (rpotter12) [reply src/packagedcode/go_mod.py:127]: I thought there can be multiple spaces also. So I did it like this.

**Comment by rpotter12** [comment]
@pombredanne Please review the PR again. My latest changes are more clear then previous ones :)

**Comment by rpotter12** [comment]
@pombredanne Please review the PR again. My latest changes are more clear then previous ones :)

**Comment by rpotter12** [comment]
In https://github.com/nexB/scancode-toolkit/pull/2230/files#diff-081f69678b5dd083e0c798c6e4d24645R126 we should use a+ but it doesn't work for me. 

I think it will work for all the cases because r'(?P<type>[^\s]+)(\s)+(?P<ns_name>[^\s]+)\s?(?P<version>(.*))' in this \s? is used because module line is not having version number so version will be optional in that case and require and exclude line (?P<version>(.*)) will detect all the characters till end.

**Review by pombredanne — APPROVED** [review]

**Comment by pombredanne** [comment]
Thanks!

**Pull request merged**